Aoc 2018 day5 - part 1, 2

Minkyu-Jeon opened the pull request.

Minkyu-Jeon committed:
aoc 2018 day5 - part1

Minkyu-Jeon committed:
aoc 2018 day5 - part2

MincheolC commented on src/aoc2018_5.clj:
```diff
@@ -0,0 +1,32 @@
+                (slurp)
+                (map int)))
+
+(defn abs [n] (if (< n 0)
```
_"nit:"_ 이전에 구글링하다 알게된 건데 아래처럼도 가능하더라고요 :) 꽤나 인상이 깊어서 공유합니다.
```clj
(fn [v] (max v (- v)))
```

Minkyu-Jeon replied:
오! 기발하네요 ㅋㅋㅋㅋ
저는 클로저 기본 함수로 제공하는 abs 함수가 있을거라고 생각했는데 못찾았어요 ㅠㅠ

tlonist-sang replied:
(Math/abs _number) 가 있긴합니다 ㅎㅎ

MincheolC commented on src/aoc2018_5.clj:
```diff
@@ -0,0 +1,32 @@
+(defn reacted? [x y] (= (abs (- x y)) 32))
+
+(defn react-polymer [state]
+  (loop [result []
```
상현님이 주신 팁인데 **reduce**로도 구현이 가능합니다! 스택을 처음부터 쓰시다니.. 대단하시네요 ㅜ_ㅠ 전 스택 생각 못해서 힘들게 풀었...는데 ㅋㅋㅋ 
```clj
(defn react-polymer [state]
  (reduce (fn [acc v]
            (if (empty? acc)
              (conj acc v)
              (if-let [_ (reacted? (peek acc) v)]
                (pop acc)
                (conj acc v))))
          [] state))
```

Minkyu-Jeon replied:
아! 훨씬 간결하군요! 저도 처음에 막 이상한 방법으로 풀다가 중간에 바꿨어요 😅

MincheolC commented on src/aoc2018_5.clj:
```diff
@@ -0,0 +1,32 @@
+
+(def input (->> "input/day5_input.txt"
+                (slurp)
+                (map int)))
```
숫자로 변환하면 훨씬 빠르네요 굿굿 👍

MincheolC commented on src/aoc2018_5.clj:
```diff
@@ -25,8 +25,30 @@
+                   (map (fn [item] (into #{} item)))))
+
+(defn remove-chars [polymer char-set]
+  (->> polymer
```
_"nit:"_ Threading macro를 안 써도 될 것 같아요!
`(filter #((complement contains?) char-set %) polymer)`

MincheolC commented on src/aoc2018_5.clj:
```diff
@@ -25,8 +25,30 @@
+
+;; part1
 (comment (->> input
               (react-polymer)
```
_"nit:"_ 저희 회사는 인자가 없는 경우 `()`을 쓰지 않는다고 하네요!

Minkyu-Jeon replied:
참고하겠습니다! 감사합니다!

tlonist-sang commented on src/aoc2018_5.clj:
```diff
@@ -0,0 +1,54 @@
+            (recur (pop result) rest)
+            (recur (conj result first-item) rest)))
+        (recur (conj result first-item) rest))
+      (conj result first-item))))
```
위에 민철님이 말하신것처럼 reduce로 O(n)으로 리팩토링하시는걸 추천드립니다!
loop-recur은 현업에서는 쓰지말자 주의에요 ㅎㅎ

Minkyu-Jeon replied:
아하 그렇군요! 감사합니닷!

tlonist-sang commented on src/aoc2018_5.clj:
```diff
@@ -0,0 +1,54 @@
+              (count)))
+
+;; part2
+(comment (->> char-vec
```
char-vec이 어떤 일을 하는 함수인지 쉽게 알 수 있도록 1)함수 이름을 구체적으로 적거나 2)용례를 코멘트로 남겨주시면 더 좋을 것 같습니다! 👍 
그리고 캐릭터를 하나씩 a부터 z까지 지우는거야말로 for를 활용해볼 수 있을 것 같아요. for가 sequence를 받아 sequence를 반환하니까, 
```
(for [i (range (int \a) (int \z)]
...
```
이런식으로 받고, body에서 input으로 받은 캐릭터 벡터 혹은 시퀀스에서 전부 지운버전을 반환하는거죠! 
이렇게 하면 좋은 점은 뒤에 로직이 깔끔해진다는 것입니다.
```
(->> (generate-seq-with-removed-chars ..)
         (map (fn [x] (react-polymer x))
         (map count)
         (apply min)         
```
문제가 요구하는 서사가 좀 더 잘 드러나면, 더 선언적이고 읽기 쉬운 코드가 될 수 있을 꺼 같습니다 :)

Minkyu-Jeon replied:
코드가 훨씬 선언적이고 읽기 쉬워지네요! 리뷰 주신대로 수정해보겠습니다!
감사합니다!!

Minkyu-Jeon committed:
refactoring: day5

- loop-recur 를 reduce로 수정
- 불필요한 Threading macro 제거
- 인자가 없는 함수에 괄호 제거
- part2 답변 구하는 부분 보다 선언적인 코드로 수정
